bleak/__init__.py

```diff
@@ -204,6 +206,55 @@ def set_scanning_filter(self, **kwargs):
         )
         self._backend.set_scanning_filter(**kwargs)
 
+    def __aiter__(self) -> AsyncIterator[Tuple[BLEDevice, AdvertisementData]]:
+        self._backend.new_devices = asyncio.Queue()
+        return self
+
+    async def __anext__(self) -> Tuple[BLEDevice, AdvertisementData]:
+        return await self._backend.new_devices.get()
+
+    @overload
+    def devices(
+        self, timeout: Optional[float] = None, return_adv: Literal[False] = False
+    ) -> AsyncIterator[BLEDevice]:
+        ...
+
+    @overload
+    def devices(
+        self, timeout: Optional[float] = None, return_adv: Literal[True] = True
+    ) -> AsyncIterator[Tuple[BLEDevice, AdvertisementData]]:
+        ...
+
+    async def devices(self, timeout: Optional[float] = None, return_adv: bool = False):
```

Comment:
I think this should be a static method similar to `discover()` (I like how `find_device_by_filter()` can now be implemented in terms of this iterator.).

Reply:
There are some filters that can actually be sent to the Bluetooth hardware, so I think we need to include options like `service_uuids: List[str]` here. Also `scanning_mode` and the various backend-specific args.

This can be implemented as `**kwargs` and just pass everything to the `BleakScanner` constructor and have the documentation refer to that.

Reply:
I thought about this before opening a PR. I think there are two ways to put it:

1. Have class methods which instantiate class instance and use it for scanning:
```python
# Pseudo code
    @classmethod
    async def discover(cls, **kwargs) -> BLEDevice:
        with cls(**kwargs) as scanner:
            for device in scanner.scan():
                yield device

async for device in BleakScanner.discover(my_filters):
    print(device)
```

2. Have instance methods and instantiate the scanner first
```python
# Pseudo code
    async def discover(self, **kwargs) -> BLEDevice:
        for device in self.scan():
            yield device

async for device in BleakScanner(my_filters).discover():
    print(device)
```

I know we cannot change from method 1 for compatibility. However, now we need a copy of docstring and typing for `**kwargs` and other scanning parameters at `BleakScan`'s `__init__()`, `set_scanning_filter()` (soon not anymore), `discover()`, `find_device_by_address()`, `find_device_by_name()`, `find_device_by_filter()`, and `scan()` or `watch_advertisements()`.

In method 2, we only require them at one place, the `__init__()` (and the scanner instance can even be reused for some extra use case).
And because I "hate" having just arbitrary `**kwargs` instead of all properly typed and documented, I made the main scan iterator method instance method, delegating the task of properly documenting all parameters to the `__init__()` method. On top of this, I would call `__init__()` in the static method anyway, just that I would need to pass all arguments.

I have now, as example, for review implemented both.
I kept the original method named `devices()` and added a `note` in the docstring - this method does not actually do anything, it only returns the devices detected - almost exactly like the property `discovered_devices`, just in the real-time and in an async way.
The class method version is called `scan()`, because it actually instantiates the class, starts scanning, and then uses `devices()` to return the devices (method 1 vs 2 again). For `kwargs` I just copied the docstring from `find_device_by_filter()`, because they mean the same.

This is just a reference update and my thought, comments welcome.

Reply:
Deleted class method as per https://github.com/hbldh/bleak/pull/1361#discussion_r1262777118